### PR TITLE
Posts List: Add Copy link functionality for posts

### LINKED ISSUE #13292: Post List: consider adding an option to copy the post URL
We've received a request in beta tester feedback to have an option to copy the URL for a post:

> The app is utter waste without the option to copy the url of published post. Its completely useless.., there is share icon and copy option, but in some phones that copy option will not appear, so we need to share it to some person in whatsapp and copy it from there.

* 2-star beta tester feedback
* Oct 27, 2020, 06:39
* Device: LGE LG G7 ThinQ
* Device language: English
* App version code: 939
* App version name: 16.0-rc-2
* Android version (SDK): Android 10 (SDK 29)

<sup>(internal reference: play-store-reviewId=gp%3AAOqpTOGxPytP9Xd-OvKfM-hn5mtQsQmTYXVx9GJjKOoDScX0q-KYXQTJvyJ0t8loRo6ydmeQYe6e8Idp3_GGDGji&corpus=BETA_FEEDBACK)</sup>

### PR DESCRIPTION
**Pull request by ovitrif**
Fixes #13292

## Description
This PR implements the "Copy link" functionality for published posts (already exists on Calypso).

We've received a request in beta tester feedback to have an option to copy the URL for a post:
>The app is utter waste without the option to copy the url of published post. Its completely useless.., there is share icon and copy option, but in some phones that copy option will not appear, so we need to share it to some person in whatsapp and copy it from there.

The proposed solution is inspired from the web version, where:
- The option to copy the post link is right above the trash button
- A notification is shown to the user after the link was copied

<details>
  <summary>Preview web version</summary>
  
  |   Copy link Button  |   Notification    |
  | ---- | ---- |
  |   1_preview_web_version  |   2_preview_web_version_notification  |
</details>

### Published/Scheduled/Draft Posts
1. From My Site press the Posts button or scroll down and select Blog Posts from the list
2. In post cards mode press the More button / In post list mode press the dots at the right of the post
3. **Verify** that the _Copy link_ button is visible above the _Trash_ button
4. Press the _Copy link_ button
5. **Verify** that a toast message appears saying "Link copied to clipboard" (see screenshots below)
6. Open a browsing app
7. Long press and paste in the address bar and navigate to that address
8. **Verify** that the link in the address bar is correct and the post is shown

### Trashed Posts
1. From My Site press the Posts button or scroll down and select Blog Posts from the list
2. Select the Trashed (4th) tab
3. Make sure you have a trashed post
2. In post cards mode press the More button / In post list mode press the dots at the right of the post
3. **Verify** that there is no _Copy link_ button

## Screenshots
| _Copy link_ Shown for Published Post | Toast Notification | _No Copy link_ shown for draft, trashed or scheduled posts |
| ---- | ---- | ---- |
|  copy_link_btn |  copy_link_toast  | scheduled_no_copy_link  |

## Regression Notes
1. Potential unintended areas of impact
Buttons for non-published posts

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing
Unit testing

3. What automated tests I added (or what prevented me from doing so)
  n/a

PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt
Comment (ovitrif) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt]: @antonis I was thinking of using this extension for all buttons which are under a simple condition captured by a boolean variable.
WDYT?

File: RELEASE-NOTES.txt
Comment (antonis) [inline RELEASE-NOTES.txt:5]: Thank you for thinking of our users and adding a release note on this @ovitrif 👍 
Adding this comment as a reminder to move the release not to the next release since `18.7` is cut today.

File: WordPress/src/main/java/org/wordpress/android/util/ContextExtensions.kt
Comment (antonis) [inline WordPress/src/main/java/org/wordpress/android/util/ContextExtensions.kt:51]: That's a good addition to the `ContextExtensions` 👍

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt
Comment (antonis) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:403]: Thank you for improving this part of the code 🙇

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt
Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt]: Great work leveraging our [Coding style tools ](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/docs/coding-style.md) and detecting the `LongMethod` issue 👍 
Though I'm a big fan of kotlin extensions I'm a bit skeptical on this since it applies just on the specific usecase/method and is not a generic extension that is discoverable and can be used in other parts of the project 🤔 
What are your thoughts on extracting some of the functionality of the `createButtonTypes` method to private functions to reduce the complexity. For example we could extract[ the last `when` statement](https://github.com/wordpress-mobile/WordPress-Android/pull/15570/files#diff-2238863dfadc135bc1b6359a21a855541d2dcf3d8cd2742d2fa91a81ea03e701L429) that adds the proper delete/trash button. Wdyt?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt
Comment (antonis) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt]: This seems like an edge case but what do you think of adding an optional call instead of a force unwrap on this and handle the case where the `clipboardManager` is null accordingly?

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt
Comment (antonis) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt:112]: I wonder if there is a way to avoid the generic exception here (thought we use it in a few other places in the code when working with the clipboard) 🤔 
The official [documentation](https://developer.android.com/guide/topics/text/copy-paste) does not cover this either. Is there a specific flow that causes an exception we can document with a comment to justify the addition of the [baseline exception](https://github.com/wordpress-mobile/WordPress-Android/pull/15570/files#diff-fa877df4f6878e67c74a6d479124a0bc5342b194c75568b95269934118273a42R352)?

**Review by antonis — COMMENTED** [review]
Great work @ovitrif 👍 
I've tested the modified functionality and it works great. The code and the tests also looks good 🎉 
I've added a few comments/suggestions but nothing crucial :)

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt
Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt]: Very good suggestion, I'm more in favor of it than what's currently in the code 🙇🏻‍♂️

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt
Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt:112]: I will research this more deeply!

To be honest at first glance I thought wrapping these calls in a try/catch is all about the NPE that could be thrown when clipboardManager is null, and I explored the option of catching a `NullPointerException` instead but of course we'll get the same complaint from Detekt - `TooGenericExceptionCaught`.

Yet there may be something more to it as I could grasp from a few searches, hope the internet will help more :)!

If a NPE is the only concern your suggestion will already improve the code, since we'll handle the null case properly!
If it's true that we may encounter other exceptions ideally we know what and document it in the code 👍 

Thanks a lot 🙇🏻‍♂️!!!

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt
Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:403]: My pleasure! 😇

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt
Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt]: Thank you, that was indeed the constraint here!

I can resonate with your skepticism as I also don't see it as something discoverable and reusable in other parts of the code, it has a pretty narrow scope.

Already did some thinking while working on the task and also saw this suggestion (to extract the last when statement for the proper delete/trash button) as a possible improvement!

I'm going to explore this suggestion and which other functionalities can be extracted while still maintaining a good overview of the buttons that are added!

File: RELEASE-NOTES.txt
Comment (ovitrif) [reply RELEASE-NOTES.txt:5]: Thank you for the remark! I noticed @AliSoftware usually updates the release notes with the new sections, would that mean it's better not to do it myself or I can proceed to add the new section?

Comment (AliSoftware) [reply RELEASE-NOTES.txt:5]: Yep, we add the new release notes section title right after we do the Code Freeze on start of each sprint then merge the new `release/*` branch we just cut back into `develop` for branch synchronization.

So indeed you usually don't have to create those new section titles yourself.

In practice, when you created your PR before the code freeze, and then the code freeze happened and added the new section for the new version after your branch was cut, I recommend you merge `develop` back into your PR branch at that point to:
 - get the new section title in `RELEASE-NOTES.txt`… allowing you to either then add the entry for your PR under it, or  move the entry you previously added to the `RELEASE-NOTES.txt` file (if you already did so like in the present case) under the new one instead;
 - get any new stuff from `develop` that might have moved on since (thus reducing the risk of potential conflicts between your branch and develop)

_In contrast, trying to add the new section yourself in your PR will just likely generate a git conflict, with 2 identical new `18.8` lines not completely in sync potentially confusing git… thus not the best option_

Comment (AliSoftware) [reply RELEASE-NOTES.txt:5]: For reference, my PR adding this new section title (as part of the code freeze sync) is [here](https://github.com/wordpress-mobile/WordPress-Android/pull/15574/files#diff-558ce8820a544da19f7fccdb2d1d661e46d68c130a1a39abdd222aa3484a5873R3) (and still pending review and not yet landed in `develop` as I write those lines)

Comment (ovitrif) [reply RELEASE-NOTES.txt:5]: Many thanks @AliSoftware for the detailed insights into this procedure 🙇🏻‍♂️

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt
Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt:112]: While researching the possible exceptions I found:

- [This article from 2013](https://commonsware.com/blog/2013/08/08/developer-psa-please-fix-your-clipboard-handling.html) which points to [an issue in Android 4.3](https://issuetracker.google.com/issues/36976027) users were experiencing having apps crash when another app / service would listen to clipboard changes.
  Since we don't support anything below Android 7.0 I think this shouldn't be a concern.
- [This chromium issue from 2020](https://bugs.chromium.org/p/chromium/issues/detail?id=1123727) which mentions possible exceptions on certain devices.

Looking more closely into [the chromium PR to fix the reported bug](https://chromium.googlesource.com/chromium/src/+/542cb9cfcc927295615809b0c99917b09a219d9f/ui/android/java/src/org/chromium/ui/base/Clipboard.java#332) we can get more insights into the possible issue and their solution:
<details>
  <summary>chromium/src/ui/android/java/src/org/chromium/ui/base/Clipboard.java:332</summary>

```java
    private boolean setPrimaryClipNoException(ClipData clip) {
        final String manufacturer = Build.MANUFACTURER.toLowerCase(Locale.US);
        // See crbug.com/1123727, there are OEM devices having strict mode violations in their
        // Android framework code. Disabling strict mode for non-google devices.
        try (StrictModeContext ignored = manufacturer.equals("google")
                ? null
                : StrictModeContext.allowAllThreadPolicies()) {
            mClipboardManager.setPrimaryClip(clip);
            return true;
        } catch (Exception ex) {
            // Ignore any exceptions here as certain devices have bugs and will fail.
            showCopyToClipboardFailureMessage();
            return false;
        }
    }
```
</details>

By looking at the [blame for that code](https://chromium.googlesource.com/chromium/src/+blame/542cb9cfcc927295615809b0c99917b09a219d9f/ui/android/java/src/org/chromium/ui/base/Clipboard.java#332) we can also notice that the call was being wrapped into a try/catch block at least since 2013-12 and there was a comment inside the `catch` stating:
```
// Ignore any exceptions here as certain devices have bugs and will fail.
```

That's all I could find on the topic In a Monday evening's research 🙃.

I'd add the above comment in the `catch` block since it may be a possible scenario but I'm also open for feedback 🙇🏻‍♂️

File: RELEASE-NOTES.txt
Comment (ovitrif) [reply RELEASE-NOTES.txt:5]: Seeing that the pull-request was merged I updated the change to release-notes

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt
Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt]: Applied the suggestion and throwing an own `NPE` if clipboard manager was null.
Please review!

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt
Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt]: Updated this too!
Thank you for the suggestions

File: RELEASE-NOTES.txt
Comment (antonis) [reply RELEASE-NOTES.txt:5]: Thank you for the change @ovitrif 🙇

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt
Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt]: Thank you for applying the suggestion with 60c19e5eea0343ebe3b80446484f1a5eea3007bf. LGTM :)

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt
Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt:112]: Thank you for your research on this @ovitrif. The resources you found and especially the chromium PR would be great for future reference 🙇 
[The comment you added](https://github.com/wordpress-mobile/WordPress-Android/pull/15570/commits/7398ab33b76a3177288e56f7d8a932530458ea55#diff-b88b70cdd11730f84fd0e57469231b42b2823575b31e6600c79783e19e0378f0R112) seems adequate to me 👍

File: WordPress/src/test/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelperTest.kt
Comment (antonis) [inline WordPress/src/test/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelperTest.kt:249]: Thank you for adding this extra test case @ovitrif 👍

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt
Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt]: Thank you for tackling this @ovitrif 🙇 
The changes introduced with 8fac44dd0f21a840733f3790fd5029423ad75a5a, 6e3550efebf5c1f6fad6bddf3fca9f86a191b1c0, 6551903b6666d97db9ba1e718ae759325340c4a9 and 48b7432792a3a4c7ff6a717d6002b0456c1266ad LGTM 🎉

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt
Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt:112]: Thank you @antonis for looking into this!
I've added [a comment with the link to the chromium changeset](https://github.com/wordpress-mobile/WordPress-Android/pull/15570/commits/6ce9767b29a03b57e7cf602cae9af0bdb8b22c5d) for future reference 🙇🏻‍♂️

**Comment by ovitrif** [comment]
Many thanks @antonis for your detailed and careful review, I'm making this pull-request ready for review and no longer a draft, and removing the [DO NOT MERGE] tag.

We still need a designer's point of view from @iamthomasbishop on the solution for the [issue](https://github.com/wordpress-mobile/WordPress-Android/issues/13292) we're addressing here 🙇🏻‍♂️

Feedback is always welcome!

**Comment by iamthomasbishop** [comment]
@ovitrif This looks great, I just have a couple of concerns:

> Verify that there is no Copy link button

Maybe I'm missing context because I haven't seen all of the previous discussion—is there a reason we're not showing the `copy link` option on "Draft" and "Scheduled" posts/pages? I took a look at the web UI, which does show it on posts/pages in "Draft" and "Scheduled" (in addition to Published) states.

> Verify that the Copy link button is visible above the Trash button

Would it be possible to add a [divider](https://material.io/components/menus#anatomy) above the "Trashed" menu option? (Note: This could be tackled separately because it's a pre-existing concern, not a result of the work here)

> Toast Notification

I think we've moved away from using Toasts in favor of [Snackbars](https://material.io/components/snackbars) for this type of notice, so that would be preferred here.

On a related note, I would also implement this as a Snackbar on iOS (which is a custom component IIRC; named either "Snackbar" or "Notice") if that's in scope—if not, would you mind creating a sibling implementation ticket over on the [WPiOS repo](https://github.com/wordpress-mobile/WordPress-iOS) and we can triage it in the near future?

**Comment by antonis** [comment]
Thank you for your feedback @iamthomasbishop 🙏

> we're not showing the `copy link` option on "Draft" and "Scheduled" posts/pages? I took a look at the web UI, which does show it on posts/pages in "Draft" and "Scheduled" (in addition to Published) states.

I missed that Thomas. I think it would be good to align with the web on this if it is possible 👍

> Would it be possible to add a [divider](https://material.io/components/menus#anatomy) above the "Trashed" menu option? (Note: This could be tackled separately because it's a pre-existing concern, not a result of the work here)

I agree that this can be tackled on a different PR later.

> I think we've moved away from using Toasts in favor of [Snackbars](https://material.io/components/snackbars) for this type of notice, so that would be preferred here.

Good catch 👍

> On a related note, I would also implement this as a Snackbar on iOS (which is a custom component IIRC; named either "Snackbar" or "Notice") if that's in scope—if not, would you mind creating a sibling implementation ticket over on the [WPiOS repo](https://github.com/wordpress-mobile/WordPress-iOS) and we can triage it in the near future?

I'll take a note to create a ticket on this to tackle separately.

**Comment by ovitrif** [comment]
Hi @iamthomasbishop, thanks a lot for the feedback 🙇🏻‍♂️!
Also many thanks to @antonis for the follow up!

> Maybe I'm missing context because I haven't seen all of the previous discussion—is there a reason we're not showing the `copy link` option on "Draft" and "Scheduled" posts/pages? I took a look at the web UI, which does show it on posts/pages in "Draft" and "Scheduled" (in addition to Published) states.

It is true that we show it also for posts/pages in "Draft" and "Scheduled" states on the web, we can definitely align the app in this regard!

> Would it be possible to add a [divider](https://material.io/components/menus#anatomy) above the "Trashed" menu option? (Note: This could be tackled separately because it's a pre-existing concern, not a result of the work here)

Definitely a nice idea! I agree that we can handle the divider in a separate task.

> I think we've moved away from using Toasts in favor of [Snackbars](https://material.io/components/snackbars) for this type of notice, so that would be preferred here.

Thank you for the information, with that in mind indeed a snackbar should be the preferred solution here.

> On a related note, I would also implement this as a Snackbar on iOS (which is a custom component IIRC; named either "Snackbar" or "Notice") if that's in scope—if not, would you mind creating a sibling implementation ticket over on the [WPiOS repo](https://github.com/wordpress-mobile/WordPress-iOS) and we can triage it in the near future?

Noted! @antonis will take care of creating the ticket for the [WPiOS](https://github.com/wordpress-mobile/WordPress-iOS)  repo.

File: WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt
Comment (ovitrif) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:384]: @antonis I checked in the wp-calypso source code and [there](https://github.com/Automattic/wp-calypso/blob/c2a3f57266c1bb6123cd9b474491c31f42fca6a9/client/my-sites/post-type-list/post-actions-ellipsis-menu/index.jsx#L21) we show the copy link button regardless of the post status.
For the mobile apps it wouldn't make sense to show it for local drafts, since this type of post will not have an URL.

I also kept it hidden for trashed posts but we can review this, since on the web version we show it:

<details>
<summary>
Web: Copy link for trashed posts</summary>

copy_link_web_trashed

</details>

Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemUiStateHelper.kt:384]: Thank you for tackling this @ovitrif 🙇 

> For the mobile apps it wouldn't make sense to show it for local drafts

This makes sense 👍 

>I also kept it hidden for trashed posts but we can review this, since on the web version we show it

I think it makes sense to keep it as is since even on the web the copied link leads to a "page not found" error. We can revisit this in the future if there is a need to add the functionality for the trashed posts.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt
Comment (antonis) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt]: Thank you for the changes @ovitrif 🙇 
On a first pass everything looks good to me apart from a minor indentation issue (not related with the latest commits) in the comment that causes a `ktlint` error:
`Unexpected indentation (17) (it should be 16)`

Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt]: Thank you @antonis! I will fix this 🙇🏻‍♂️

Comment (ovitrif) [reply WordPress/src/main/java/org/wordpress/android/ui/posts/PostListAction.kt]: Checked with `ktlint`, fixed the formatting and ran `ktlint` again, now it's successful!

Merged with the latest develop branch too

**Review by antonis — APPROVED** [review]
Great work @ovitrif 👍 
I retested the changes with the latest commits and everything works as expected 🎉 
The code also LGTM 🚢

**Pull request merged**
